Fix timestamp on the timeseries based charts (UTC -> LOCAL TZ)

shmsr opened the pull request:
The previous behaviour on the time series based charts was that the time was represented in UTC which wasn't user-friendly.

After the fix, the timestamp is shown in the client's local timezone which is taken from the browser of the client.

Note: It also solves the problem of hive query failing for date-time columns (when in epoch_ms format) as hive doesn't support `/` operator rather it supports `DIV`

### CATEGORY

Choose one

- [x] Bug Fix
- [x] Enhancement (new features, refinement)
- [ ] Refactor
- [ ] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

Screenshots are corresponding to `bintime = 1569906000` and `Timezone = Asia/Calcutta`

From epoch converter, results are:
`UTC = Tuesday, October 1, 2019 5:00:00 AM`
`IST = Tuesday, October 1, 2019 10:30:00 AM GMT+05:30`

We can hence verify the values from the chart where we have hovered on.

### Before:

Chart:
<img width="1440" alt="Before 1" src="https://user-images.githubusercontent.com/51480165/68697674-56123d00-05a5-11ea-9fad-6c0b9a72ba23.png">

Query:
<img width="1440" alt="Before 2" src="https://user-images.githubusercontent.com/51480165/68697697-62969580-05a5-11ea-9919-084012de331a.png">

### After:

Chart:
<img width="1440" alt="After 1" src="https://user-images.githubusercontent.com/51480165/68697712-6aeed080-05a5-11ea-9487-8f059b8519c4.png">

Query:
<img width="1440" alt="After 2" src="https://user-images.githubusercontent.com/51480165/68697718-6d512a80-05a5-11ea-93ea-6be5b21162b3.png">

### TEST PLAN
<!--- What steps should be taken to verify the changes -->

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### REVIEWERS

shmsr committed:
Fix timestamp on the timeseries based charts issue:

> Rather than UTC, timestamps of charts are in client's TZ
> Now user's have right representation of charts

subham-sarkar-guavus commented on superset/templates/appbuilder/footer.html:
```diff
@@ -3,7 +3,7 @@
 <div class="navbar navbar-default navbar-fixed-bottom navbar-template navbar-default-template navbar-fixed-bottom-template">
     <div class="footer-left">
-      <label class='footer-left-label'>{{ timezone }}</label>
+      <label class='footer-left-label'>SERVER TZ: {{ timezone }}</label>
```
Yes. Suggest me the name, I'll do it. Timezone mentioned there was full of conflict, whether it's the timezone where the Superset server is hosted or the timezone of the client from where superset is accessed. So added the prefix there.

subham-sarkar-guavus replied:
Writing UTC every time is very misleading. Suppose if the server's in `America/New York` timezone, and for the calculations, it'll pick that. Although, I have pointed out changes to make it UTC always, in another PR. So, the differences of the relativeness for `epoch_s` and `epoch_ms` timestamps are taken relative from UTC which is correct for all further calculations. I've also verified the data manually from hive, and the output's satisfactory.

So, writing UTC is misleading client-side users.

arpit-agarwal replied:
Is this actually a server timezone?
If yes i would say `Application Timezone`.
However in context of other PR it become irrelevant.
Assuming both will be merged together we can discuss what should be suffix

subham-sarkar-guavus commented on superset/db_engine_specs.py:
```diff
@@ -115,16 +115,20 @@ class BaseEngineSpec(object):
         # if epoch, translate to DATE using db specific conf
         if pdf == 'epoch_s':
-            expr = cls.epoch_to_dttm().format(col=expr)
+            sub_expr = 'from_utc_timestamp(from_unixtime(%s), \'%s\')' % (expr, timezone)
```
Worked for most of the SQL based DBs (Maybe all).
For example, works with, MySQL, Hive, MariaDB, Postgres, etc.

Ref: https://www.tutorialspoint.com/sql/sql-date-functions.htm

You can find all the functions I've used here, so it's not hive specifically.

Plus, consider using `DIV` and not `\` also helps as when the format of unix_timestamp is in epoch_ms we divide it by 1000 and then do the query. So using `DIV` works for all given the fact that `\` fails for Hive. So, after making the changes, we're good to go for `epoch_ms` formats as well.

subham-sarkar-guavus replied:
Also, prior to this, the `epoch_to_dttm` used to return this `from_unixtime(unix_timestamp(from_unixtime(bin_time)), 'yyyy-MM-dd HH')`, I'm just using this one, `from_unixtime(unix_timestamp(from_utc_timestamp(from_unixtime(bin_time), 'Asia/Calcutta')), 'yyyy-MM-dd HH')`. Hence, no issues.

subham-sarkar-guavus replied:
I've resolved some issues that could've broken this code. Also, as per the suggested I'm using the usual `epoch_to_dttm()` only. Hope the changes look good. They also passed all the UTs.

shmsr committed:
Fix UTs & code for chart timestamp's timezone specific changes

shmsr committed:
Add the epoch_s/ epoch_ms column

As the column is converted to local timezone, data validation might be harder from actual datasource and UI
Now due to present of column in epoch_s or epoch_ms, one can validate the data

arpit-agarwal commented on superset/connectors/sqla/models.py:
```diff
@@ -163,8 +163,12 @@ def get_timestamp_expression(self, time_grain):
+            grain,
+            timezone)
+        # Add the date-time column in select statement for data validation (epoch_s/ epoch_ms)
+        return [literal_column(expr, type_=DateTime).label(label), literal_column(self.expression or self.column_name)]
```
Is this relevant to changes for showing client timezone in timeseries

arpit-agarwal reviewed (changes requested):
I need to understand overall flow and scope of this PR.
This seems to doing lot more than just timeseries time in local timezone

arpit-agarwal commented on superset/db_engine_specs.py:
```diff
@@ -1480,7 +1487,7 @@ class PinotEngineSpec(BaseEngineSpec):
 
     @classmethod
-    def get_time_expr(cls, expr, pdf, time_grain, grain):
+    def get_time_expr(cls, expr, pdf, time_grain, grain, timezone):
```
why we have added a redundant parameter

arpit-agarwal commented on superset/viz.py:
```diff
@@ -333,6 +337,7 @@ def query_obj(self):
         }
 
         d = {
+            'timezone': client_tz,
```
where this timzone property is used.
can you share a flow of method invocations that is stitching this e2e data flow

arpit-agarwal reviewed (commented):
#DYNAMIC TZ
`select to_utc_timestamp(from_unixtime(1389802875), from_unixtime(unix_timestamp(), "z"));`

We may need this is we don't know the timezone at which `from_unixtime` will convert to
